clients/src/main/java/org/apache/kafka/common/network/ListenerName.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import org.apache.kafka.common.protocol.SecurityProtocol;
+
+import java.util.Locale;
+import java.util.Objects;
+
+public final class ListenerName {
+
+    /**
+     * Create an instance with the security protocol name as the value.
+     */
+    public static ListenerName forSecurityProtocol(SecurityProtocol securityProtocol) {
+        return new ListenerName(securityProtocol.name);
+    }
+
+    /**
+     * Create an instance with the provided value converted to uppercase.
+     */
+    public static ListenerName normalised(String value) {
+        return new ListenerName(value.toUpperCase(Locale.ROOT));
+    }
+
+    private final String value;
+
+    public ListenerName(String value) {
```

Comment:
I was wondering about that. I decided to keep it like this because there are a number of situations where we know that the String has already been normalised, so we don't need to do it again. However, always normalising is safer so I am willing to change it if you feel strongly.